src/plexil-adapter/OwInterface.cpp

```diff
@@ -746,140 +646,235 @@ void OwInterface::takePicture (int id)
+  action_thread.detach();
+}
 
+void OwInterface::grindAction (double x, double y, double depth, double length,
+                               bool parallel, double ground_pos, int id)
+{
+  ow_lander::GrindGoal goal;
+  goal.x_start = x;
+  goal.y_start = y;
+  goal.depth = depth;
+  goal.length = length;
+  goal.parallel = parallel;
+  goal.ground_position = ground_pos;
 
-  if (check_service_client (client, Op_Unstow)) {
-    ow_lander::Unstow srv;
-    thread service_thread (call_ros_service<ow_lander::Unstow>,
-                           client, srv, Op_Unstow, id);
-    service_thread.detach();
-  }
+  runAction<Grind, actionlib::SimpleActionClient<ow_lander::GrindAction>,
+            ow_lander::GrindGoal,
+            ow_lander::GrindResultConstPtr,
+            ow_lander::GrindFeedbackConstPtr>
+    (Op_Grind, m_grindClient, goal, id);
 }
 
-void OwInterface::deliverSample (double x, double y, double z, int id)
+void OwInterface::guardedMove (double x, double y, double z,
+                               double dir_x, double dir_y, double dir_z,
+                               double search_dist, int id)
 {
-  if (! mark_operation_running (Op_DeliverSample, id)) return;
+  if (! mark_operation_running (Op_GuardedMove, id)) return;
+  thread action_thread (&OwInterface::guardedMoveAction, this, x, y, z,
+                        dir_x, dir_y, dir_z, search_dist, id);
+  action_thread.detach();
+}
 
-  ros::NodeHandle nhandle ("planning");
+void OwInterface::guardedMoveAction (double x, double y, double z,
+                                     double dir_x, double dir_y, double dir_z,
+                                     double search_dist, int id)
+{
+  ow_lander::GuardedMoveGoal goal;
+  goal.start.x = x;
+  goal.start.y = y;
+  goal.start.z = z;
+  goal.normal.x = dir_x;
+  goal.normal.y = dir_y;
+  goal.normal.z = dir_z;
+  goal.search_distance = search_dist;
 
-  ros::ServiceClient client =
-    nhandle.serviceClient<ow_lander::DeliverSample>("/arm/deliver_sample");
+  // NOTE: this function does not use runAction() typically called here, because
+  // of the specialized 'done' callback.
 
+  thread fault_thread (monitor_for_faults, Op_GuardedMove);
 
-  if (check_service_client (client, Op_DeliverSample)) {
-    ow_lander::DeliverSample srv;
-    srv.request.x = x;
-    srv.request.y = y;
-    srv.request.z = z;
-    thread service_thread (call_ros_service<ow_lander::DeliverSample>,
-                           client, srv, Op_DeliverSample, id);
-    service_thread.detach();
+  if (m_guardedMoveClient) {
+    m_guardedMoveClient->sendGoal
+      (goal,
+       guarded_move_done_cb,
+       active_cb<GuardedMove>,
+       action_feedback_cb<ow_lander::GuardedMoveFeedbackConstPtr>);
   }
+  else {
+    ROS_ERROR ("m_guardedMoveClient was null!");
+    return;
+  }
+
+  // Wait for the action to return
+  bool finished_before_timeout =
+    m_guardedMoveClient->waitForResult (ros::Duration (ActionTimeoutSecs));
```

Comment:
Serious problem: GuardedMove (and all actions) block for this timeout duration, regardless of how fast they complete.

Reply:
@Samahu suggests this may be a ROS issue.  waitForResult should not block for the timeout duration if the action finishes earlier.  And I have not consistently reproduced this problem.   Considering it out of scope of this review.